### PR TITLE
Fix DtoServer error messages

### PR DESCRIPTION
**Pull request by dmytro-ndp**
### What does this PR do?
Fix error messages in case of there is no implementation of DTO interface available for DtoServer.

@evoevodin, @sleshchenko, @skabashnyuk: please, take a look at this request.

### CODE REVIEW

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java [hunk 1]
```diff
@@ -386,8 +386,13 @@ public Gson getGson() {
     private <T> DtoProvider<T> getDtoProvider(Class<T> dtoInterface) {
         DtoProvider<?> dtoProvider = dtoInterface2Providers.get(dtoInterface);
         if (dtoProvider == null) {
-            throw new IllegalArgumentException("Unknown DTO type " + dtoInterface);
+            if (dtoInterface.isAnnotationPresent(DTO.class)) {
+                throw new IllegalArgumentException("Implementation for DTO type " + dtoInterface + " isn't found.");
```
Comment (voievodin) [inline core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: Maybe it is better to say _"Provider for DTO type ..."_ ?

**Review by voievodin — APPROVED** [review]

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java (hunk 1 above)
Comment (dmytro-ndp) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: Thank you for advice.
The main reason of this request is to mention **implementation** in some way in error message.
But in the method which throws exception we operate with term **provider**.
Next message is enclosing both of reasons:
> Provider of implementation for DTO type ...

Comment (voievodin) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: A good one, go with it

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java [hunk 2]
```diff
@@ -386,8 +386,13 @@ public Gson getGson() {
     private <T> DtoProvider<T> getDtoProvider(Class<T> dtoInterface) {
         DtoProvider<?> dtoProvider = dtoInterface2Providers.get(dtoInterface);
         if (dtoProvider == null) {
-            throw new IllegalArgumentException("Unknown DTO type " + dtoInterface);
+            if (dtoInterface.isAnnotationPresent(DTO.class)) {
+                throw new IllegalArgumentException("Provider of implementation for DTO type " + dtoInterface + " isn't found");
+            } else {
+                throw new IllegalArgumentException("Is not a DTO type " + dtoInterface);
```
Comment (skabashnyuk) [inline core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: what if "blah-blah" is not a DTO type
?

Comment (dmytro-ndp) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: Then there will be an IllegalArgumentException:  "Is not a DTO type class blah-blah", won't it?

Comment (sleshchenko) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: I think Sergii means what if rewrite message in following way: `dtoInterface + " is not a DTO type"`

**Review by sleshchenko — APPROVED** [review]
It will be much easier to understand a problem with lack of dto implementations after your changes. :+1:

**Review by skabashnyuk — APPROVED** [review]

**Comment by dmytro-ndp** [comment]
@skabashnyuk, @evoevodin, @sleshchenko: please, review my new commit with next fixes:
1. Handle non-interface types.
1. Check on DTO annotation among the parent interfaces as well.
1. Fix of error message due to @skabashnyuk suggestion.

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java [hunk 3]
```diff
@@ -386,12 +388,37 @@ public Gson getGson() {
     private <T> DtoProvider<T> getDtoProvider(Class<T> dtoInterface) {
         DtoProvider<?> dtoProvider = dtoInterface2Providers.get(dtoInterface);
         if (dtoProvider == null) {
-            throw new IllegalArgumentException("Unknown DTO type " + dtoInterface);
+            if (!dtoInterface.isInterface()) {
+                throw new IllegalArgumentException(format("Only interfaces can be DTO, but %s is not", dtoInterface));
+            }
+
+            if (checkOnDtoAnnotation(dtoInterface)) {
```
Comment (sleshchenko) [inline core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: as for me `check...` method should throw an exception and `is...` or `has...` should return true or false.

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java [hunk 4]
```diff
@@ -386,12 +388,37 @@ public Gson getGson() {
...
         }
+
         return (DtoProvider<T>)dtoProvider;
     }
 
     /**
+     * Checks if dtoInterface or its parent have DTO annotation.
+     * @param dtoInterface
+     *          DTO interface
+     * @return true if onle dtoInterface or one of its parent have DTO annotation.
```
Comment (sleshchenko) [inline core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: typo `onle`

Comment (dmytro-ndp) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: Thank you!

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java (hunk 3 above)
Comment (dmytro-ndp) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: I see. Let's rename it to **hasDtoAnnotation**.

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java [hunk 5]
```diff
@@ -386,12 +388,37 @@ public Gson getGson() {
...
+     */
+    private boolean checkOnDtoAnnotation(Class dtoInterface) {
+        for (Class parent: dtoInterface.getInterfaces()) {
+            if (checkOnDtoAnnotation(parent)) {
+                return true;
+            }
+        }
+
+        return dtoInterface.isAnnotationPresent(DTO.class);
```
Comment (voievodin) [inline core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: I think it's better to check it before analyzing interfaces, for example:
```java
@DTO
interface A extends B, C { }
```

the code above will analyze B and C before analyzing A, which i think unreasonable.

Comment (dmytro-ndp) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java]: Good advice! Thank you!

**Review by skabashnyuk — APPROVED** [review]

File: core/che-core-api-dto/src/test/java/org/eclipse/che/dto/definitions/DTOHierarchy.java [hunk 6]
```diff
@@ -60,4 +62,6 @@
         void setShadowedField(GrandchildDto v);
     }
 
+    public interface GrandchildWithoutDto extends Child, Closeable {
```
Comment (sleshchenko) [inline core/che-core-api-dto/src/test/java/org/eclipse/che/dto/definitions/DTOHierarchy.java]: Why should it be Closeable? It can confuse =)

Comment (sleshchenko) [reply core/che-core-api-dto/src/test/java/org/eclipse/che/dto/definitions/DTOHierarchy.java]: Also Mb you are able to reuse just `Child` class or create `Grandchild` class if you want three level of extending

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java [hunk 7]
```diff
@@ -386,12 +388,41 @@ public Gson getGson() {
...
+            } else {
+                throw new IllegalArgumentException(dtoInterface + " is not a DTO type");
+            }
         }
+
         return (DtoProvider<T>)dtoProvider;
     }
 
     /**
+     * Checks if dtoInterface or its parent have DTO annotation.
```
Comment (sleshchenko) [inline core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java:406]: It would be good to add an empty line after summary section.

**Review by sleshchenko — APPROVED** [review]

File: core/che-core-api-dto/src/test/java/org/eclipse/che/dto/definitions/DTOHierarchy.java (hunk 6 above)
Comment (dmytro-ndp) [reply core/che-core-api-dto/src/test/java/org/eclipse/che/dto/definitions/DTOHierarchy.java]: Closeable is just additional interface for testing propose. I am ready to produce second child if you  think it's less confusing :-)

File: core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java (hunk 7 above)
Comment (dmytro-ndp) [reply core/che-core-api-dto/src/main/java/org/eclipse/che/dto/server/DtoFactory.java:406]: I see. Will fix it.
Thank you for note!

File: core/che-core-api-dto/src/test/java/org/eclipse/che/dto/definitions/DTOHierarchy.java (hunk 6 above)
Comment (sleshchenko) [reply core/che-core-api-dto/src/test/java/org/eclipse/che/dto/definitions/DTOHierarchy.java]: Up to you

**Pull request merged**